keps/sig-node/3610-env-injection/README.md

```diff
@@ -0,0 +1,920 @@
+
+2. The second proposal is using a new API Object named GlobalEnv like `LimitRange` or `ResourceQuota`.
+
+- In the spec, we can define a list of env using `EnvVar`, so that we can support both `value` and `valueFrom` in the spec.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The idea came from <https://github.com/kubernetes/kubernetes/issues/48180>.
+
+1. Indicate what environment or location we are in, e.g. ENVIRONMENT=dev or ENVIRONMENT=prod. 
+The cluster administrator should be able to say, "define the following environment vars that are injected into every container in the cluster (or, at best, limited by namespace).
+
+2. Enabling features. E.g.
+I might want APM enabled on many of my app containers, but only when run in a kube cluster (as opposed to just running the image on my laptop), or perhaps only in certain environments. Cluster manager should be able to set a global env var of ENABLE_APM=true that then causes any container that consumes it to know, "I must enable APM now."
+
+3. Inject namespace into env
+Automatically inject the kubernetes namespace into an environment variable for me. With dozens of different kube templates people sometimes forget to expose it manually (<https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/>)
+
+4. Have compute nodes have special knowledge about themselves that can be injected into environment variables. We have a odd/even compute node concept for upgrading openshift and this information may come in handy
+
+5. Be able to specify system wide environment variables. We have multiple openshift instances (different environments and different datacenters). It would be great if that data could be made automatically available without having to deal with custom env sections in configs.
+
+6. `http_proxy`: We have a proxy in our datacenter and we need to set `http_proxy` for all containers. <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-356866173>
+
+7. `KUBERNETES_SERVICE_HOST`: Being able to set KUBERNETES_SERVICE_HOST cluster wide makes it possible to use an external hostname (with a publicly signed SSL certificate) without any issues. <https://github.com/kubernetes/kubernetes/issues/48180#issuecomment-375111483>
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- if the env is not set in container, inject env to all containers in namespace
+- if it is already set in container, override some env and not override for others
```

Comment:
If one updated the custom resource `GlobalEnv` on a cluster with custom values in a given namespace, will that forcefully inject the updated values to the pods on the cluster again or one has to delete the pod manually and force the injection of these env? If the latter, then should we have that under `non-goal`?

Reply:
Ppd updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds`, `spec.tolerations`

For most other plugins, the injection only happens once the pod is created. We can add this to non-goal.

Reply:
> Ppd updates may not change fields other than spec.containers[*].image, spec.initContainers[*].image, spec.activeDeadlineSeconds, spec.tolerations

Agreed. The only way to update the envs would be to trigger a delete of the pod. Which is definitely not the right thing to do. A line under non-goal would be good enough I think. Thanks